### PR TITLE
Upgrade to Storybook 7

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
This PR upgrades Storybook to Storybook 7, along with a host of other related upgrades. 
- Upgrades React to 18
- Removes storyshot and instead uses Jest snapshots in storybook-test-runner. This uses a hosted storybook and takes snapshots of the rendered DOM, so the rest of the changes focus on making sure the storybooks render the same on any computer.

First of all, dates. We can't rely on current time or the computer's timezone to render stories, since those vary.

- Updates to a custom version of storybook-react-intl. (PR open soon to upstream my changes). This custom version allows specifying the Timezone to react-intl, so it is the same across computers and in CI.
- Adds a new Storybook extension to mock the JS new Date(). In relevant stories (ItineraryBody) we now assign a fixed date.

There was also an issue with the elevation profile view, which uses an auto resize component to grab the page width. For some reason this was rendering differently in CI, likely due to differences between the headless browser used on Linux by Playwright and on MacOS. To solve this, I check the user agent and when we are in Playwright, we can mock this value and fix the width to a constant. This is in `leg-diagram-preview.tsx`.

AXE tests have also changed. [The latest way to do this](https://storybook.js.org/blog/automate-accessibility-tests-with-storybook/) is with AXE-Playwright, which just runs in the Playwright tests. These tests have detected many more errors than before, which is causing the tests to fail. I don't know if these are false positives or if we had tests disabled before, or if the new checker has more rules.

The final issue that I still don't really know the cause of was that we had build problems when the default export of a component was different from the file name itself. We had many icons where the name of the icon began with Svg, while the filename did not. For example, `export default SvgCar` in `Car.js`. Webpack 5 (the new build system with Storybook 7). I fixed this by renaming the exports in all these files. I wish I had been able to find why this wasn't working, since I believe it should be. 

For future reference, this is how I accomplished this rename:
```
❯ find . -type f -execdir sed -ibak "s/Svg[^ ;]*/$(basename {} .js)/" {} \;  
❯ find . -type f -execdir sed -ibak "s/\.js//" {} \;                                                                                  
❯ rm *bak
```

### CODE REVIEW

File: packages/building-blocks/tsconfig.json [hunk 1]
```diff
@@ -8,6 +8,5 @@
     "lib": ["es2019", "dom"],
     "skipLibCheck": true
   },
-  "include": ["src/**/*"],
-  "exclude": ["src/__tests__/**/*"]
```
Comment (daniel-heppner-ibigroup) [inline packages/building-blocks/tsconfig.json:12]: This exclude was overriding the exclude from the extended config at the root level

File: packages/core-utils/tsconfig.json [hunk 2]
```diff
@@ -8,6 +8,5 @@
     "lib": ["es2019", "dom"],
     "skipLibCheck": true
   },
-  "include": ["src/**/*"],
-  "exclude": ["src/__tests__/**/*"]
```
Comment (daniel-heppner-ibigroup) [inline packages/core-utils/tsconfig.json:12]: This exclude was overriding the exclude from the extended config at the root level

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Some cleanup to do but mostly this is very impressive! Fantastic improvement and cleanly implemented. 

Only bug is that `yarn dev` seems not to work right. Growing pains?
Screenshot 2024-05-12 at 4 58 00 PM

File: packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.tsx [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import React from "react";
 import { action } from "@storybook/addon-actions";
-import { boolean } from "@storybook/addon-knobs";
+// import { boolean } from "@storybook/addon-knobs";
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/vehicle-rental-overlay/src/VehicleRentalOverlay.story.tsx]: Yes! I needed to add the controls. Got that done

File: packages/otp2-tile-overlay/src/mocks/handlers.js [hunk 4]
```diff
@@ -3,9 +3,12 @@ import { rest } from "msw";
...
 
-import seventy from "!url-loader!./4770-6206.pbf";
-import seventyOne from "!url-loader!./4771-6206.pbf";
-import seventyTwo from "!url-loader!./4772-6206.pbf";
+// import seventy from "./4770-6206.pbf";
+// import seventyOne from "./4771-6206.pbf";
+// import seventyTwo from "./4772-6206.pbf";
+const seventy = new URL("./4770-6206.pbf", import.meta.url);
+const seventyOne = new URL("./4771-6206.pbf", import.meta.url);
+const seventyTwo = new URL("./4772-6206.pbf", import.meta.url);
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/otp2-tile-overlay/src/mocks/handlers.js:11]: I left the comments because I was scared

[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/otp2-tile-overlay/src/mocks/handlers.js:11]: gonezo now

File: snapshot-serializer.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// The jest-serializer-html package is available as a dependency of the test-runner
+const jestSerializerHtml = require("jest-serializer-html");
+
+const DYNAMIC_ID_PATTERN = /"\b[A-Fa-f0-9]{8}-[A-Fa-f0-9]{4}-[A-Fa-f0-9]{4}-[A-Fa-f0-9]{4}-[A-Fa-f0-9]{12}\b"/g;
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply snapshot-serializer.js:4]: no, it's referenced in `test-runner-config.js` so we can put it wherever. Maybe in .storybook?

[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply snapshot-serializer.js:4]: Yeah I like that!

File: .storybook/react-intl.ts [hunk 6]
```diff
@@ -22,11 +22,9 @@ const packages = [
   "location-field",
   "printable-itinerary",
   "map-popup",
-  "stops-overlay",
   "transit-vehicle-overlay",
   "trip-details",
-  "trip-form",
-  "vehicle-rental-overlay"
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply .storybook/react-intl.ts:29]: bump

**Comment by miles-grant-ibigroup** [comment]
Still seeing this error: 
image
Other than that thanks for the changes

File: README.md [hunk 7]
```diff
@@ -37,6 +37,8 @@ Some packages in otp-ui depend on sibling packages (e.g., `@opentripplanner/core
 2. Run:
    `yarn && yarn dev`
 
+If the Storybook addon bar (a bar of controls at the bottom of the story) does not appear, you may need to clear localStorage by opening the browser console and typing `localStorage.clear()`.
```
Comment (miles-grant-ibigroup) [inline README.md:40]: For me I had to run `yarn clean`

**Review by miles-grant-ibigroup — APPROVED** [review]
I've been testing, I've been trying to break it... But I think this is a win! But we should be careful about merging. Lots of upstream changes that we need to merge in here carefully

File: README.md (hunk 7 above)
Comment (daniel-heppner-ibigroup) [reply README.md:40]: storybook bug that's been around for years

**Comment by miles-grant-ibigroup** [comment]
a11y tests are failing. Sounds like `otp2-tile-layer` doesn't like the a11y renderer

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Crucial change before I can approve again

File: packages/transit-vehicle-overlay/src/index.story.tsx [hunk 8]
```diff
@@ -52,14 +54,6 @@ export const OuterCaretWithCustomSize = () => (
...
 );
 
-export const RouteColorBackground = () => (
-  <TransitVehicleOverlay
-    IconContainer={withRouteColorBackground(DefaultIconContainer)}
-    ModeIcon={TriMetModeIcon}
-    vehicles={vehicles}
-  />
-);
-
```
Comment (miles-grant-ibigroup) [inline packages/transit-vehicle-overlay/src/index.story.tsx:62]: Why was this removed

**Review by miles-grant-ibigroup — APPROVED** [review]
Really excited about this. Great work

File: packages/transit-vehicle-overlay/src/index.story.tsx (hunk 8 above)
Comment (daniel-heppner-ibigroup) [reply packages/transit-vehicle-overlay/src/index.story.tsx:62]: bad merge maybe?

File: packages/base-map/src/index.story.tsx [hunk 9]
```diff
@@ -24,7 +24,8 @@ const center: [number, number] = [45.522862, -122.667837];
 export default {
   args: { center },
   component: BaseMap,
-  title: "BaseMap"
+  title: "BaseMap",
+  parameters: { storyshots: { disable: true } }
```
Comment (binh-dam-ibigroup) [inline packages/base-map/src/index.story.tsx:28]: Move variable `disableStoryShots` to above this declaration and use it here... Do we need to explicity add it to the other stories? Same thing in the other packages.

File: packages/icons/src/leg-icon.js [hunk 10]
```diff
@@ -10,11 +10,9 @@ const LegIcon = ({ getCompanyIcon, leg, ModeIcon, ...props }) => {
   if (company && typeof getCompanyIcon === "function") {
     const CompanyIcon = getCompanyIcon(company);
     if (CompanyIcon)
-      return (
-        <Suspense fallback={<span>{company}</span>}>
-          <CompanyIcon {...props} />
-        </Suspense>
-      );
+      <SafeSuspense fallback={<span>{company}</span>}>
```
Comment (binh-dam-ibigroup) [inline packages/icons/src/leg-icon.js:13]: Note that this is for accommodating tests only, and that no release will be triggered for this package.

File: packages/map-popup/package.json [hunk 11]
```diff
@@ -11,15 +11,15 @@
   "license": "MIT",
   "private": false,
   "dependencies": {
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/map-popup/package.json:14]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/trip-details/package.json [hunk 12]
```diff
@@ -11,7 +11,7 @@
   "license": "MIT",
   "private": false,
   "dependencies": {
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/trip-details/package.json:14]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/core-utils/src/index.ts [hunk 13]
```diff
@@ -22,4 +23,5 @@ const core = {
   ui
 };
 
+export { SafeSuspense };
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/index.ts:26]: Note that this is to accommodate tests run locally only, and that no release will be triggered for this package.

File: packages/trip-form/package.json [hunk 14]
```diff
@@ -10,7 +10,7 @@
   "types": "lib/index.d.ts",
   "private": false,
   "dependencies": {
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/package.json:13]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/zoom-based-markers/package.json [hunk 15]
```diff
@@ -19,14 +19,14 @@
     "url": "https://github.com/opentripplanner/otp-ui/issues"
   },
   "dependencies": {
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/zoom-based-markers/package.json:22]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/vehicle-rental-overlay/package.json [hunk 16]
```diff
@@ -20,7 +20,7 @@
   },
   "dependencies": {
     "@opentripplanner/base-map": "^3.1.0",
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/vehicle-rental-overlay/package.json:23]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/trip-viewer-overlay/package.json [hunk 17]
```diff
@@ -21,10 +21,10 @@
   "dependencies": {
     "@mapbox/polyline": "^1.1.0",
     "@opentripplanner/base-map": "^3.1.0",
-    "@opentripplanner/core-utils": "^11.4.0"
+    "@opentripplanner/core-utils": "^11.4.1"
```
Comment (binh-dam-ibigroup) [inline packages/trip-viewer-overlay/package.json:24]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/trip-form/src/util.tsx [hunk 18]
```diff
@@ -269,9 +270,9 @@ function getTransitCombinedModeOptions(
             selectedCompanies.includes(modeCompanyUpper)),
         text: (
           <span>
-            <Suspense fallback={company || "Loading..."}>
+            <SafeSuspense fallback={company || "Loading..."}>
```
Comment (binh-dam-ibigroup) [inline packages/trip-form/src/util.tsx:273]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/transit-vehicle-overlay/package.json [hunk 19]
```diff
@@ -10,15 +10,15 @@
   "private": false,
   "dependencies": {
     "@opentripplanner/base-map": "^3.1.0",
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/transit-vehicle-overlay/package.json:13]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/stop-viewer-overlay/package.json [hunk 20]
```diff
@@ -20,13 +20,13 @@
   },
   "dependencies": {
     "@opentripplanner/base-map": "^3.1.0",
-    "@opentripplanner/core-utils": "^11.4.0"
+    "@opentripplanner/core-utils": "^11.4.1"
```
Comment (binh-dam-ibigroup) [inline packages/stop-viewer-overlay/package.json:23]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/route-viewer-overlay/package.json [hunk 21]
```diff
@@ -21,15 +21,15 @@
   "dependencies": {
     "@mapbox/polyline": "^1.1.0",
     "@opentripplanner/base-map": "^3.1.0",
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/route-viewer-overlay/package.json:24]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/location-field/src/index.tsx [hunk 22]
```diff
@@ -954,7 +954,7 @@ const LocationField = ({
       </S.HiddenContent>
       <ItemList
         // Hide the list from screen readers if no enabled options are shown.
-        aria-hidden={hasNoEnabledOptions}
+        aria-hidden={hasNoEnabledOptions || !menuVisible}
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:957]: I noted the `fix`commit for this change, which will trigger a release for this package.

File: packages/itinerary-body/package.json [hunk 23]
```diff
@@ -10,7 +10,7 @@
   "license": "MIT",
   "private": false,
   "dependencies": {
-    "@opentripplanner/core-utils": "^11.4.0",
+    "@opentripplanner/core-utils": "^11.4.1",
```
Comment (binh-dam-ibigroup) [inline packages/itinerary-body/package.json:13]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

File: packages/itinerary-body/src/AccessLegBody/leg-diagram-preview.tsx [hunk 24]
```diff
@@ -121,7 +124,7 @@ class LegDiagramPreview extends Component<Props, State> {
             </S.PreviewDiagramElevationLoss>
           </S.PreviewDiagramTitle>
           {profile.points.length > 0 ? (
-            generateSvg(profile, width)
+            generateSvg(profile, IS_TEST_RUNNER ? "245" : width)
```
Comment (binh-dam-ibigroup) [inline packages/itinerary-body/src/AccessLegBody/leg-diagram-preview.tsx:127]: Note that this is to accommodate tests only, and that no release will be triggered for this package.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Could you check consistency in using the story parameters to disable storyshots? Sometimes they are applied component wide, other times they are applied for each story.

File: package.json [hunk 25]
```diff
@@ -70,18 +71,19 @@
...
     "lerna": "^3.18.4",
     "lint-staged": "^8.2.0",
-    "msw": "^0.34.0",
+    "msw": "^2.3.1",
     "nock": "^11.7.0",
     "prettier": "^1.19.1",
     "puppeteer": "^10.2.0",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
+    "react": "^18.2.0",
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply package.json:78]: I think so, good point. I'll get that PR ready.

**Comment by daniel-heppner-ibigroup** [comment]
Thanks for the feedback on consistency of disabling storyshots. I have moved all of them into the default export and got rid of the disableStoryshots variables since they're only used in one place now.

File: packages/endpoints-overlay/src/EndpointsOverlay.story.tsx [hunk 26]
```diff
@@ -43,7 +41,8 @@ function CatDogIcon({ type }: UserLocationAndType) {
 export default {
   component: EndpointsOverlay,
   decorators: [withMap()],
-  title: "EndpointsOverlay"
+  title: "EndpointsOverlay",
+  parameters: { storyshots: { disable: true } }
```
Comment (binh-dam-ibigroup) [inline packages/endpoints-overlay/src/EndpointsOverlay.story.tsx:45]: Nit: Sort attributes

**Review by binh-dam-ibigroup — APPROVED** [review]
Going ahead and approving. That's a needed upgrade.

File: packages/endpoints-overlay/src/EndpointsOverlay.story.tsx (hunk 26 above)
Comment (daniel-heppner-ibigroup) [reply packages/endpoints-overlay/src/EndpointsOverlay.story.tsx:45]: oh oops, I forgot to do this.
